### DISCUSSION TITLE
Threashold for residual

### DISCUSSION THREAD

**Discussion by ashishdhole** [idaholab/moose#28247] [Q&A Modules: Phase field]
Hello,
I am working on grain growth model with particles. In my simulations, grain growth stops after certain time and reach a steady state. Although I could see the grain g rowth rate to be reducing with time, They never reach a steady state. instead, dt keeps reducing until the point when the simulation terminates with an error that dt_min is reached. is there a way I can define in my code the minimum threadhold beyond which the residual becomes zero? 
Regards, 
Ashish

**Comment by GiudGiud** [comment]
Hello

This is not the normal progression of the simulation. The solver is failing which is why the time step keep getting reduced. 

Please see our "troubleshooting failed solves" instructions. 
The first step is to identify whether the linear or nonlinear solve is failing

Guillaume

**Reply by ashishdhole** [reply]
what should be my next step after identifying the fail?
I am certain that non-linear solve is failing. 

following is what my executioner block looks like
```
[Executioner]
  type = Transient
  scheme = bdf2
  solve_type = PJFNK #Preconditioned JFNK (default)
  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre    boomeramg'
  steady_state_detection = true
  steady_state_start_time = 0
  steady_state_tolerance = 1e-8
  l_abs_tol = 1.0e-3
  nl_abs_tol = 1.0e-3
  l_tol = 1.0e-3
  l_max_its = 5
  nl_max_its = 5
  nl_rel_tol = 1.0e-3
  start_time = 0.0
  #num_steps = 200
  end_time = 20000000000
  dtmin = 1e-20
  #dt = 1
  line_search = none

  [Adaptivity]
    initial_adaptivity = 0
    refine_fraction = 0.7
    coarsen_fraction = 0.1
    max_h_level = 1
  []
  [TimeStepper]
    type = IterationAdaptiveDT
    cutback_factor = 0.8
    dt = 0.00001
    growth_factor = 1.2
    optimal_iterations = 7
  []
[]
```
Can you suggest optimum values for these?

Ashish

**Reply by GiudGiud** [reply]
can you paste the simulation log here?

**Reply by ashishdhole** [reply]
Here is the output file

[0p05.txt](https://github.com/user-attachments/files/16432111/0p05.txt)

**Reply by GiudGiud** [reply]
Thé tolerances are super loose

Please tighten the nonlinear convergence criterion. Accepting non converged solutions is not helping the solver finish the transient

**Reply by ashishdhole** [reply]
```
  l_abs_tol = 1.0e-4
  nl_abs_tol = 1.0e-4
  l_tol = 1.0e-4
  l_max_its = 30
  nl_max_its = 30
  nl_rel_tol = 1.0e-4
```
will that work. If I go less than this the solve takes really long time. I tried with tight tolerances earlier, but, for every timesteps it was taking such a huge time.

**Reply by GiudGiud** [reply]
There s no way to know ahead of time if it will be enough. 

You can look at improving preconditioning to make the linear solves faster and be able to afford more nonlinear solves. 

You should use checkpoitining to make a checkpoint before the problem arises, then tighten the convergence tolerances a few time steps before it occurred
